src/main/java/com/spotify/futures/ConcurrencyLimiter.java

```diff
@@ -120,18 +120,16 @@ public int remainingActiveCapacity() {
    * 2) return null and does not acquire a permit from the semaphore
    */
   private Job<T> grabJob() {
-    while (!queue.isEmpty()) {
-      if (!limit.tryAcquire()) {
-        return null;
-      }
-
-      final Job<T> job = queue.poll();
-      if (job != null) {
-        return job;
-      }
+    if (!limit.tryAcquire()) {
+      return null;
+    }
 
-      limit.release();
+    final Job<T> job = queue.poll();
+    if (job != null) {
+      return job;
     }
+
+    limit.release();
```

Comment:
should be wrapped in `finally` because you are not releasing it after the return statement on line 129.